### PR TITLE
Add reduce_scatter_tensor to CommTensor

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90681
* #90569
* #90565
* __->__ #90564
* #90512

This PR adds reduce_scatter_base to the CommTensor

### CODE REVIEW

**Comment by mrshenli** [comment]
XLA failure look irrelevant

```
2022-12-09T23:04:17.8551222Z ======================================================================
2022-12-09T23:04:17.8551331Z ERROR: test_get_xla_tensor (__main__.TestSelect)
2022-12-09T23:04:17.8551537Z ----------------------------------------------------------------------
2022-12-09T23:04:17.8551628Z Traceback (most recent call last):
2022-12-09T23:04:17.8551791Z   File "/var/lib/jenkins/workspace/xla/test/test_operations.py", line 666, in test_get_xla_tensor
2022-12-09T23:04:17.8551863Z     t = x.data.cpu()
2022-12-09T23:04:17.8552119Z RuntimeError: /var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor.cpp:187 : Check failed: handle->HasValue() 
2022-12-09T23:04:17.8552196Z *** Begin stack trace ***
2022-12-09T23:04:17.8552295Z 	tsl::CurrentStackTrace[abi:cxx11]()
2022-12-09T23:04:17.8552388Z 	torch_xla::XLATensor::GetXlaData()
2022-12-09T23:04:17.8552481Z 	torch_xla::XLATensor::ToTensor(bool)
2022-12-09T23:04:17.8552638Z 	torch_xla::XLANativeFunctions::_copy_from(at::Tensor const&, at::Tensor const&, bool)
2022-12-09T23:04:17.8552701Z 	
2022-12-09T23:04:17.8552829Z 	at::_ops::_copy_from::call(at::Tensor const&, at::Tensor const&, bool)
2022-12-09T23:04:17.8552875Z 	
2022-12-09T23:04:17.8552989Z 	at::native::copy_(at::Tensor&, at::Tensor const&, bool)
2022-12-09T23:04:17.8553109Z 	at::_ops::copy_::call(at::Tensor&, at::Tensor const&, bool)
2022-12-09T23:04:17.8553343Z 	at::native::_to_copy(at::Tensor const&, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, bool, c10::optional<c10::MemoryFormat>)
2022-12-09T23:04:17.8553402Z 	
2022-12-09T23:04:17.8553666Z 	at::_ops::_to_copy::redispatch(c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, bool, c10::optional<c10::MemoryFormat>)
2022-12-09T23:04:17.8553725Z 	
2022-12-09T23:04:17.8553980Z 	at::_ops::_to_copy::redispatch(c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, bool, c10::optional<c10::MemoryFormat>)
2022-12-09T23:04:17.8554045Z 	
2022-12-09T23:04:17.8554278Z 	at::_ops::_to_copy::call(at::Tensor const&, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, bool, c10::optional<c10::MemoryFormat>)
2022-12-09T23:04:17.8554507Z 	at::native::to(at::Tensor const&, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, bool, bool, c10::optional<c10::MemoryFormat>)
2022-12-09T23:04:17.8554555Z 	
2022-12-09T23:04:17.8554800Z 	at::_ops::to_dtype_layout::call(at::Tensor const&, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, bool, bool, c10::optional<c10::MemoryFormat>)
2022-12-09T23:04:17.8554859Z 	
2022-12-09T23:04:17.8554920Z 	
2022-12-09T23:04:17.8554977Z 	
2022-12-09T23:04:17.8555070Z 	_PyMethodDescr_FastCallKeywords
2022-12-09T23:04:17.8555164Z 	
2022-12-09T23:04:17.8555234Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8555322Z 	_PyFunction_FastCallKeywords
2022-12-09T23:04:17.8555403Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8555482Z 	_PyEval_EvalCodeWithName
2022-12-09T23:04:17.8555588Z 	_PyObject_FastCallDict
2022-12-09T23:04:17.8555647Z 	
2022-12-09T23:04:17.8555717Z 	PyObject_Call
2022-12-09T23:04:17.8555784Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8555864Z 	_PyEval_EvalCodeWithName
2022-12-09T23:04:17.8556060Z 	_PyObject_FastCallDict
2022-12-09T23:04:17.8556117Z 	
2022-12-09T23:04:17.8556364Z 	_PyObject_FastCallKeywords
2022-12-09T23:04:17.8556630Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8556928Z 	_PyEval_EvalCodeWithName
2022-12-09T23:04:17.8557077Z 	_PyObject_FastCallDict
2022-12-09T23:04:17.8557136Z 	
2022-12-09T23:04:17.8557251Z 	PyObject_Call
2022-12-09T23:04:17.8557440Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8557607Z 	_PyEval_EvalCodeWithName
2022-12-09T23:04:17.8557929Z 	_PyObject_FastCallDict
2022-12-09T23:04:17.8558056Z 	
2022-12-09T23:04:17.8558300Z 	_PyObject_FastCallKeywords
2022-12-09T23:04:17.8558424Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8558547Z 	_PyEval_EvalCodeWithName
2022-12-09T23:04:17.8558706Z 	_PyObject_FastCallDict
2022-12-09T23:04:17.8558933Z 	
2022-12-09T23:04:17.8559311Z 	PyObject_Call
2022-12-09T23:04:17.8559503Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8559640Z 	_PyEval_EvalCodeWithName
2022-12-09T23:04:17.8559716Z 	_PyObject_FastCallDict
2022-12-09T23:04:17.8559802Z 	
2022-12-09T23:04:17.8559994Z 	_PyObject_FastCallKeywords
2022-12-09T23:04:17.8560185Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8560361Z 	_PyFunction_FastCallKeywords
2022-12-09T23:04:17.8560605Z 	
2022-12-09T23:04:17.8560719Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8560847Z 	_PyFunction_FastCallKeywords
2022-12-09T23:04:17.8560928Z 	
2022-12-09T23:04:17.8561049Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8561182Z 	_PyEval_EvalCodeWithName
2022-12-09T23:04:17.8561276Z 	_PyObject_FastCallDict
2022-12-09T23:04:17.8561389Z 	
2022-12-09T23:04:17.8561471Z 	_PyObject_FastCallKeywords
2022-12-09T23:04:17.8561555Z 	
2022-12-09T23:04:17.8561695Z 	_PyEval_EvalFrameDefault
2022-12-09T23:04:17.8561789Z 	_PyEval_EvalCodeWithName
2022-12-09T23:04:17.8561879Z 	PyEval_EvalCode
2022-12-09T23:04:17.8561937Z 	
2022-12-09T23:04:17.8562047Z 	PyRun_FileExFlags
2022-12-09T23:04:17.8562172Z 	PyRun_SimpleFileExFlags
2022-12-09T23:04:17.8562260Z 	
2022-12-09T23:04:17.8562418Z 	_Py_UnixMain
2022-12-09T23:04:17.8562520Z 	__libc_start_main
2022-12-09T23:04:17.8562593Z 	
2022-12-09T23:04:17.8562692Z *** End stack trace ***
2022-12-09T23:04:17.8562862Z Trying to access XLA data while an async operation is in flight: UNKNOWN_SCALAR[]
2022-12-09T23:04:17.8562868Z 
```

**Comment by mrshenli** [comment]
BC error is real:

```
2022-12-09T21:48:31.0373002Z The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 
2022-12-09T21:48:31.0373018Z 
2022-12-09T21:48:31.0373088Z Broken ops: [
2022-12-09T21:48:31.0373514Z 	c10d::_reduce_scatter_base_(Tensor _0, Tensor _1, __torch__.torch.classes.c10d.ProcessGroup _2, __torch__.torch.classes.c10d.ReduceOp _3, int _4) -> __torch__.torch.classes.c10d.Work _0
2022-12-09T21:48:31.0373563Z ]
```

File: torch/csrc/distributed/c10d/Ops.cpp
Comment (mrshenli) [inline torch/csrc/distributed/c10d/Ops.cpp:123]: curious, when will this method be invoked and when will we dispatch to the cpu/cuda version below?

**Review by mrshenli — APPROVED** [review]
Stamp to unblock. Please address BC error before landing.

File: torch/csrc/distributed/c10d/Ops.cpp
Comment (H-Huang) [reply torch/csrc/distributed/c10d/Ops.cpp:123]: Yes, the definition in Ops.cpp is redundant and will get removed in https://github.com/pytorch/pytorch/pull/88330, feel free to land this PR as-is though!

**Pull request closed without merging**
